akka-http-core/src/main/scala/akka/http/scaladsl/UseHttp2.scala

```diff
@@ -14,4 +14,5 @@ object UseHttp2 {
   object Never extends UseHttp2
   object Negotiated extends UseHttp2
   object Always extends UseHttp2
+  object PriorKnowledge extends UseHttp2
```

Comment:
@jrudolph OTOH, is there any reason why you wouldn't want to run both HTTP/1.1 and HTTP/2 at all time?

Reply:
I suspect there might be a performance penalty to that?

Reply:
@jrudolph @raboof If we collapse this feature into Negotiated, what would that look like in practice?

Reply:
I'll give it a go

Reply:
@raboof :+1: let me know when/if you want me to review :)

Reply:
Tried that in https://github.com/akka/akka-http/pull/2553

I think a fourth option makes sense. That would make the matrix:
* `Never`: no HTTP/2 support, no overhead relative to just HTTP/1.1
* `Negotiated`: ALPN when secure, support for h2c via 'Upgrade' when unsecure, no/low overhead compared to just HTTP/1.1
* `PriorKnowledge`: n/a when secure, support for both h2c mechanism when unsecure, some overhead (from `PriorKnowledgeSwitch`)
* `Always`: no HTTP/1.1 support

Am I overestimating the performance penalty of `PriorKnowledgeSwitch`?

Reply:
I suspect that this might get even more complicated once HTTP/3 is released.
How about we just add so that the user lists what protocols they want to use? If the concern is purely a performance one, then perhaps work instead should be placed into minimizing overhead of the implementation of the features orthogonally to the configuration?

protocols = [HTTP1] == Never
protocols = [HTTP1, HTTP2] == Negotiated
protocols = [HTTP2] == Always

Reply:
I would even simplify things even more: either you disable HTTP/2 (default), so there's no performance penalty for HTTP/1.1, or you enable HTTP/2 in addition to HTTP/1.1, and you get a small performance penalty for the switches. I don't see the reason why we necessarily would have to support `Always` at all, since the only added feature over `Negotiated` is failing when someone tries to connect with HTTP/1.1 (+ the performance benefit which alone might not be worth adding a feature for it).

The benefit would be that we remove UseHttp2 completely and just use the `enable-http2` flag.

Reply:
@jrudolph Sounds good to me! Less code to maintain and less permutations to worry about.

Reply:
> The benefit would be that we remove UseHttp2 completely and just use the enable-http2 flag

Yes that sounds good, since `enable-http2` can still be tweaked via `ServerSettings` if really needed.

Reply:
@raboof @jrudolph Let's do it!